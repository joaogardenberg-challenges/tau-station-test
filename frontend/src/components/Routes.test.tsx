import { render, screen } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'

const useHistory = jest.fn()

jest.doMock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory
}))

jest.mock('./images_list/ImagesList')
jest.mock('./image_modal/ImageModal')

const Routes = require('./Routes').default

const renderWithRouter = (
  ui: any,
  { route = '/' }: { route: string } = { route: '/' },
  options: any = {}
) => {
  window.history.pushState({}, 'Router Test', route)
  return render(ui, { wrapper: BrowserRouter, ...options })
}

describe('Routes', () => {
  beforeEach(() => {
    useHistory.mockReturnValue({ replace: jest.fn() })
  })

  it('renders images list successfully', () => {
    const { container } = renderWithRouter(<Routes />, { route: '/images' })

    const imagesList = screen.getByTestId('images-list')
    const imageModal = screen.queryByTestId('image-modal')

    expect(useHistory().replace).not.toHaveBeenCalled()
    expect(imagesList).toBeInTheDocument()
    expect(imageModal).not.toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('renders image modal successfully', () => {
    const { container } = renderWithRouter(<Routes />, { route: '/images/1' })

    const imagesList = screen.getByTestId('images-list')
    const imageModal = screen.getByTestId('image-modal')

    expect(useHistory().replace).not.toHaveBeenCalled()
    expect(imagesList).toBeInTheDocument()
    expect(imageModal).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('redirects "/" to "/images"', () => {
    renderWithRouter(<Routes />)
    expect(useHistory().replace).toHaveBeenCalledTimes(1)
    expect(useHistory().replace).toHaveBeenCalledWith('/images')
  })
})
