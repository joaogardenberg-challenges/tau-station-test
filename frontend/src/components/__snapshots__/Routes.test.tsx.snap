// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routes renders image modal successfully 1`] = `
<div>
  <div
    data-testid="image-modal"
  />
  <div
    data-testid="images-list"
  />
</div>
`;

exports[`Routes renders images list successfully 1`] = `
<div>
  <div
    data-testid="images-list"
  />
</div>
`;
